### ISSUE TITLE
Connect timeout on windows when the server isn't running instead of immediate Connection Refused

### ISSUE THREAD

**Issue by andyp1per** [warmcat/libwebsockets#570]
Trying to connect to a nonexistent server works great on Linux/OSX - the connection times out almost immediately. 
On windows this doesn't happen. The connection attempt is never failed and we hit our own timeout code.
I tried setting timeout_secs in the creation info to 1s and it makes no difference.
Anything I could be doing wrong?

**Comment by andyp1per** [comment]
This is on 1.7.6 BTW

**Comment by lws-team** [comment]
Not enough info to guess...maybe the connection failed cleanly, but on a path that doesn't do the CONNECTION_ERROR callback.

Make sure you're building in debug

cmake .. -DCMAKE_BUILD_TYPE=DEBUG

and crank the logging right up (like -d65535) and see if any clue there.

**Comment by andyp1per** [comment]
Ok easy to demo
./libwebsockets-test-client.exe 127.0.0.1 --port=49151
hangs

**Comment by lws-team** [comment]
Yah... I don't have windows.  So posting the logs as requested is the way forward.

**Comment by andyp1per** [comment]
Hmmn so in the cmake command I get this report:

  Manually-specified variables were not used by the project:

    CMAKE_BUILD_TYPE

**Comment by lws-team** [comment]
That works on Linux... maybe if it's vs, you set the project mode to debug somehow... basically it defines _DEBUG and lets the low priority debug messages get compiled in.

**Comment by andyp1per** [comment]
Ok I think this is compiled in already under the Debug configuration. This is what I get from the tool:

Eagle Andy> ./libwebsockets-test-client.exe 127.0.0.1 -d65535 -p49151
[2016/05/23 15:12:51:8270] NOTICE: libwebsockets test client - license LGPL2.1+SLE
[2016/05/23 15:12:51:8270] NOTICE: (C) Copyright 2010-2016 Andy Green <andy@warmcat.com>
[2016/05/23 15:12:51:8270] NOTICE: Initial logging level 7
[2016/05/23 15:12:51:8270] NOTICE: Libwebsockets version: 1.7.7 eaglendy@Eagle-v1.7.7-1-gf89e90f
[2016/05/23 15:12:51:8270] NOTICE: IPV6 not compiled in
[2016/05/23 15:12:51:8270] NOTICE: libev support not compiled in
[2016/05/23 15:12:51:8280] NOTICE:  Threads: 1 each 30000 fds
[2016/05/23 15:12:51:8300] NOTICE:  mem: per-conn:          320 bytes + protocol rx buf
[2016/05/23 15:12:51:8360] NOTICE:  canonical_hostname = Eagle
[2016/05/23 15:12:52:0130] NOTICE:  OpenSSL doesn't support ECDH
[2016/05/23 15:12:52:0130] NOTICE: dumb: connecting
[2016/05/23 15:12:52:0150] NOTICE: mirror: connecting
[2016/05/23 15:13:13:0262] NOTICE: wsi 0242E818: TIMEDOUT WAITING on 2 (did hdr 0, ah 0079E0A0, wl 0, pfd events 0)
[2016/05/23 15:13:13:0262] ERR: mirror: LWS_CALLBACK_CLIENT_CONNECTION_ERROR
[2016/05/23 15:13:13:0262] NOTICE: lws_header_table_detach: wsi 0242E818: ah held 21s, ah.rxpos 0, ah.rxlen 0, mode/state 8 4,wsi->more_rx_waiting 0
[2016/05/23 15:13:13:0262] NOTICE: wsi 0242CD98: TIMEDOUT WAITING on 2 (did hdr 0, ah 0079D420, wl 0, pfd events 0)
[2016/05/23 15:13:13:0262] ERR: dumb: LWS_CALLBACK_CLIENT_CONNECTION_ERROR
[2016/05/23 15:13:13:0262] NOTICE: lws_header_table_detach: wsi 0242CD98: ah held 21s, ah.rxpos 0, ah.rxlen 0, mode/state 8 4,wsi->more_rx_waiting 0
[2016/05/23 15:13:13:0262] NOTICE: dumb: connecting
[2016/05/23 15:13:13:0262] NOTICE: mirror: connecting
[2016/05/23 15:13:34:0504] NOTICE: wsi 0242CD98: TIMEDOUT WAITING on 2 (did hdr 0, ah 0079E0A0, wl 0, pfd events 0)
[2016/05/23 15:13:34:0504] ERR: mirror: LWS_CALLBACK_CLIENT_CONNECTION_ERROR
[2016/05/23 15:13:34:0504] NOTICE: lws_header_table_detach: wsi 0242CD98: ah held 21s, ah.rxpos 0, ah.rxlen 0, mode/state 8 4,wsi->more_rx_waiting 0
[2016/05/23 15:13:34:0504] NOTICE: wsi 0242E818: TIMEDOUT WAITING on 2 (did hdr 0, ah 0079D420, wl 0, pfd events 0)
[2016/05/23 15:13:34:0504] ERR: dumb: LWS_CALLBACK_CLIENT_CONNECTION_ERROR
[2016/05/23 15:13:34:0504] NOTICE: lws_header_table_detach: wsi 0242E818: ah held 21s, ah.rxpos 0, ah.rxlen 0, mode/state 8 4,wsi->more_rx_waiting 0
[2016/05/23 15:13:34:0504] NOTICE: dumb: connecting
[2016/05/23 15:13:34:0504] NOTICE: mirror: connecting
[2016/05/23 15:13:55:0526] NOTICE: wsi 0242CD98: TIMEDOUT WAITING on 2 (did hdr 0, ah 0079E0A0, wl 0, pfd events 0)
[2016/05/23 15:13:55:0526] ERR: mirror: LWS_CALLBACK_CLIENT_CONNECTION_ERROR
[2016/05/23 15:13:55:0526] NOTICE: lws_header_table_detach: wsi 0242CD98: ah held 21s, ah.rxpos 0, ah.rxlen 0, mode/state 8 4,wsi->more_rx_waiting 0
[2016/05/23 15:13:55:0526] NOTICE: wsi 0242E818: TIMEDOUT WAITING on 2 (did hdr 0, ah 0079D420, wl 0, pfd events 0)
[2016/05/23 15:13:55:0526] ERR: dumb: LWS_CALLBACK_CLIENT_CONNECTION_ERROR
[2016/05/23 15:13:55:0526] NOTICE: lws_header_table_detach: wsi 0242E818: ah held 21s, ah.rxpos 0, ah.rxlen 0, mode/state 8 4,wsi->more_rx_waiting 0
[2016/05/23 15:13:55:0526] NOTICE: dumb: connecting
[2016/05/23 15:13:55:0536] NOTICE: mirror: connecting
[2016/05/23 15:14:16:0558] NOTICE: wsi 0242CD98: TIMEDOUT WAITING on 2 (did hdr 0, ah 0079E0A0, wl 0, pfd events 0)
[2016/05/23 15:14:16:0558] ERR: mirror: LWS_CALLBACK_CLIENT_CONNECTION_ERROR
[2016/05/23 15:14:16:0558] NOTICE: lws_header_table_detach: wsi 0242CD98: ah held 21s, ah.rxpos 0, ah.rxlen 0, mode/state 8 4,wsi->more_rx_waiting 0
[2016/05/23 15:14:16:0558] NOTICE: wsi 0242E818: TIMEDOUT WAITING on 2 (did hdr 0, ah 0079D420, wl 0, pfd events 0)
[2016/05/23 15:14:16:0558] ERR: dumb: LWS_CALLBACK_CLIENT_CONNECTION_ERROR
[2016/05/23 15:14:16:0558] NOTICE: lws_header_table_detach: wsi 0242E818: ah held 21s, ah.rxpos 0, ah.rxlen 0, mode/state 8 4,wsi->more_rx_waiting 0
[2016/05/23 15:14:16:0558] NOTICE: dumb: connecting
[2016/05/23 15:14:16:0568] NOTICE: mirror: connecting
[2016/05/23 15:14:37:0630] NOTICE: wsi 0242CD98: TIMEDOUT WAITING on 2 (did hdr 0, ah 0079E0A0, wl 0, pfd even

**Comment by andyp1per** [comment]
Note that there is a 21s delay between the first connecting message and the TIMEDOUT message

**Comment by lws-team** [comment]
He is timing out on PENDING_TIMEOUT_AWAITING_CONNECT_RESPONSE and firing the connection error.  I assume the timeout in the context is smaller and that's the only problem.

It feels like one of the connect apis simply blocks for 20s on your platform.

Maybe you can use the debugger to break it during the 20s and see where it is.

**Comment by andyp1per** [comment]
My console was hiding stuff. Does this tell you any more?
[2016/05/23 16:41:19:0775] NOTICE: libwebsockets test client - license LGPL2.1+S
LE
[2016/05/23 16:41:19:0775] NOTICE: (C) Copyright 2010-2016 Andy Green <andy@warm
cat.com>
[2016/05/23 16:41:19:0775] NOTICE: Initial logging level 65535
[2016/05/23 16:41:19:0775] NOTICE: Libwebsockets version: 1.7.8 eaglendy@Eagle-v
1.7.8-3-g8d4c673
[2016/05/23 16:41:19:0785] NOTICE: IPV6 not compiled in
[2016/05/23 16:41:19:0785] NOTICE: libev support not compiled in
[2016/05/23 16:41:19:0785] INFO:  LWS_DEF_HEADER_LEN    : 1024
[2016/05/23 16:41:19:0785] INFO:  LWS_MAX_PROTOCOLS     : 5
[2016/05/23 16:41:19:0785] INFO:  LWS_MAX_SMP           : 1
[2016/05/23 16:41:19:0785] INFO:  SPEC_LATEST_SUPPORTED : 13
[2016/05/23 16:41:19:0785] INFO:  sizeof (*info)        : 128
[2016/05/23 16:41:19:0785] INFO:  SYSTEM_RANDOM_FILEPATH: '/dev/urandom'
[2016/05/23 16:41:19:0795] INFO:  default timeout (secs): 20
[2016/05/23 16:41:19:0795] NOTICE:  Threads: 1 each 30000 fds
[2016/05/23 16:41:19:0805] INFO:  mem: context:          4952 bytes (856 ctx + (
1 thr x 4096))
[2016/05/23 16:41:19:0805] INFO:  mem: http hdr rsvd:   67584 bytes (1 thr x (10
24 + 3200) x 16))
[2016/05/23 16:41:19:0805] INFO:  mem: pollfd map:      240000
[2016/05/23 16:41:19:0815] INFO:  LWS_MAX_EXTENSIONS_ACTIVE: 2
[2016/05/23 16:41:19:0815] NOTICE:  mem: per-conn:          320 bytes + protocol
 rx buf
[2016/05/23 16:41:19:0875] NOTICE:  canonical_hostname = Eagle
[2016/05/23 16:41:19:2385] NOTICE:  SSL ECDH curve 'prime256v1'
[2016/05/23 16:41:19:2385] NOTICE: dumb: connecting
[2016/05/23 16:41:19:2395] INFO: lws_header_table_attach: wsi 0099CE78: ah 00000
000 (tsi 0)
[2016/05/23 16:41:19:2395] INFO: lws_header_table_attach: wsi 0099CE78: ah 004C8
3F0: count 1 (on exit)
[2016/05/23 16:41:19:2395] CLIENT: lws_client_connect: direct conn
[2016/05/23 16:41:19:2395] CLIENT: lws_client_connect_2
[2016/05/23 16:41:19:2395] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:19:2405] DEBUG: insert_wsi_socket_into_fds: 0099CE78: tsi=0, s
ock=332, pos-in-fds=0
[2016/05/23 16:41:19:2415] CLIENT: nonblocking connect retry
[2016/05/23 16:41:19:2415] NOTICE: mirror: connecting
[2016/05/23 16:41:19:2415] INFO: lws_header_table_attach: wsi 0099E648: ah 00000
000 (tsi 0)
[2016/05/23 16:41:19:2415] INFO: lws_header_table_attach: wsi 0099E648: ah 004C9
070: count 2 (on exit)
[2016/05/23 16:41:19:2425] CLIENT: lws_client_connect: direct conn
[2016/05/23 16:41:19:2425] CLIENT: lws_client_connect_2
[2016/05/23 16:41:19:2425] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:19:2425] DEBUG: insert_wsi_socket_into_fds: 0099E648: tsi=0, s
ock=380, pos-in-fds=1
[2016/05/23 16:41:19:2425] CLIENT: nonblocking connect retry
[2016/05/23 16:41:19:2425] CLIENT: lws_client_connect_2
[2016/05/23 16:41:19:2435] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:19:2435] CLIENT: nonblocking connect retry
[2016/05/23 16:41:19:2435] CLIENT: lws_client_connect_2
[2016/05/23 16:41:19:2435] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:19:2435] CLIENT: nonblocking connect retry
[2016/05/23 16:41:19:7436] CLIENT: lws_client_connect_2
[2016/05/23 16:41:19:7436] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:19:7446] CLIENT: nonblocking connect retry
[2016/05/23 16:41:19:7446] CLIENT: lws_client_connect_2
[2016/05/23 16:41:19:7456] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:19:7466] CLIENT: nonblocking connect retry
[2016/05/23 16:41:20:2476] CLIENT: lws_client_connect_2
[2016/05/23 16:41:20:2476] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:20:2486] CLIENT: nonblocking connect retry
[2016/05/23 16:41:20:2496] CLIENT: lws_client_connect_2
[2016/05/23 16:41:20:2506] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:20:2516] CLIENT: nonblocking connect retry
[2016/05/23 16:41:20:7516] CLIENT: lws_client_connect_2
[2016/05/23 16:41:20:7516] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:20:7536] CLIENT: nonblocking connect retry
[2016/05/23 16:41:20:7546] CLIENT: lws_client_connect_2
[2016/05/23 16:41:20:7556] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:20:7576] CLIENT: nonblocking connect retry
[2016/05/23 16:41:21:2587] CLIENT: lws_client_connect_2
[2016/05/23 16:41:21:2587] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:21:2597] CLIENT: nonblocking connect retry
[2016/05/23 16:41:21:2597] CLIENT: lws_client_connect_2
[2016/05/23 16:41:21:2597] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:21:2607] CLIENT: nonblocking connect retry
[2016/05/23 16:41:21:7607] CLIENT: lws_client_connect_2
[2016/05/23 16:41:21:7607] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:21:7617] CLIENT: nonblocking connect retry
[2016/05/23 16:41:21:7627] CLIENT: lws_client_connect_2
[2016/05/23 16:41:21:7637] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:21:7637] CLIENT: nonblocking connect retry
[2016/05/23 16:41:22:2647] CLIENT: lws_client_connect_2
[2016/05/23 16:41:22:2647] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:22:2667] CLIENT: nonblocking connect retry
[2016/05/23 16:41:22:2677] CLIENT: lws_client_connect_2
[2016/05/23 16:41:22:2677] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:22:2687] CLIENT: nonblocking connect retry
[2016/05/23 16:41:22:7697] CLIENT: lws_client_connect_2
[2016/05/23 16:41:22:7697] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:22:7707] CLIENT: nonblocking connect retry
[2016/05/23 16:41:22:7717] CLIENT: lws_client_connect_2
[2016/05/23 16:41:22:7727] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:22:7727] CLIENT: nonblocking connect retry
[2016/05/23 16:41:23:2738] CLIENT: lws_client_connect_2
[2016/05/23 16:41:23:2738] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:23:2748] CLIENT: nonblocking connect retry
[2016/05/23 16:41:23:2758] CLIENT: lws_client_connect_2
[2016/05/23 16:41:23:2768] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:23:2778] CLIENT: nonblocking connect retry
[2016/05/23 16:41:23:7788] CLIENT: lws_client_connect_2
[2016/05/23 16:41:23:7788] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:23:7798] CLIENT: nonblocking connect retry
[2016/05/23 16:41:23:7798] CLIENT: lws_client_connect_2
[2016/05/23 16:41:23:7808] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:23:7808] CLIENT: nonblocking connect retry
[2016/05/23 16:41:24:2818] CLIENT: lws_client_connect_2
[2016/05/23 16:41:24:2818] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:24:2828] CLIENT: nonblocking connect retry
[2016/05/23 16:41:24:2828] CLIENT: lws_client_connect_2
[2016/05/23 16:41:24:2828] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:24:2838] CLIENT: nonblocking connect retry
[2016/05/23 16:41:24:7839] CLIENT: lws_client_connect_2
[2016/05/23 16:41:24:7839] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:24:7849] CLIENT: nonblocking connect retry
[2016/05/23 16:41:24:7859] CLIENT: lws_client_connect_2
[2016/05/23 16:41:24:7869] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:24:7869] CLIENT: nonblocking connect retry
[2016/05/23 16:41:25:2879] CLIENT: lws_client_connect_2
[2016/05/23 16:41:25:2879] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:25:2879] CLIENT: nonblocking connect retry
[2016/05/23 16:41:25:2889] CLIENT: lws_client_connect_2
[2016/05/23 16:41:25:2889] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:25:2889] CLIENT: nonblocking connect retry
[2016/05/23 16:41:25:7899] CLIENT: lws_client_connect_2
[2016/05/23 16:41:25:7899] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:25:7919] CLIENT: nonblocking connect retry
[2016/05/23 16:41:25:7929] CLIENT: lws_client_connect_2
[2016/05/23 16:41:25:7939] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:25:7949] CLIENT: nonblocking connect retry
[2016/05/23 16:41:26:2959] CLIENT: lws_client_connect_2
[2016/05/23 16:41:26:2959] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:26:2959] CLIENT: nonblocking connect retry
[2016/05/23 16:41:26:2969] CLIENT: lws_client_connect_2
[2016/05/23 16:41:26:2969] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:26:2969] CLIENT: nonblocking connect retry
[2016/05/23 16:41:26:7980] CLIENT: lws_client_connect_2
[2016/05/23 16:41:26:7980] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:26:8000] CLIENT: nonblocking connect retry
[2016/05/23 16:41:26:8010] CLIENT: lws_client_connect_2
[2016/05/23 16:41:26:8020] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:26:8030] CLIENT: nonblocking connect retry
[2016/05/23 16:41:27:3040] CLIENT: lws_client_connect_2
[2016/05/23 16:41:27:3040] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:27:3040] CLIENT: nonblocking connect retry
[2016/05/23 16:41:27:3050] CLIENT: lws_client_connect_2
[2016/05/23 16:41:27:3050] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:27:3060] CLIENT: nonblocking connect retry
[2016/05/23 16:41:27:8060] CLIENT: lws_client_connect_2
[2016/05/23 16:41:27:8060] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:27:8080] CLIENT: nonblocking connect retry
[2016/05/23 16:41:27:8080] CLIENT: lws_client_connect_2
[2016/05/23 16:41:27:8090] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:27:8100] CLIENT: nonblocking connect retry
[2016/05/23 16:41:28:3111] CLIENT: lws_client_connect_2
[2016/05/23 16:41:28:3111] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:28:3121] CLIENT: nonblocking connect retry
[2016/05/23 16:41:28:3131] CLIENT: lws_client_connect_2
[2016/05/23 16:41:28:3141] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:28:3151] CLIENT: nonblocking connect retry
[2016/05/23 16:41:28:8151] CLIENT: lws_client_connect_2
[2016/05/23 16:41:28:8151] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:28:8161] CLIENT: nonblocking connect retry
[2016/05/23 16:41:28:8171] CLIENT: lws_client_connect_2
[2016/05/23 16:41:28:8181] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:28:8191] CLIENT: nonblocking connect retry
[2016/05/23 16:41:29:3191] CLIENT: lws_client_connect_2
[2016/05/23 16:41:29:3191] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:29:3191] CLIENT: nonblocking connect retry
[2016/05/23 16:41:29:3201] CLIENT: lws_client_connect_2
[2016/05/23 16:41:29:3201] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:29:3201] CLIENT: nonblocking connect retry
[2016/05/23 16:41:29:8212] CLIENT: lws_client_connect_2
[2016/05/23 16:41:29:8212] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:29:8222] CLIENT: nonblocking connect retry
[2016/05/23 16:41:29:8242] CLIENT: lws_client_connect_2
[2016/05/23 16:41:29:8242] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:29:8252] CLIENT: nonblocking connect retry
[2016/05/23 16:41:30:3262] CLIENT: lws_client_connect_2
[2016/05/23 16:41:30:3262] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:30:3262] CLIENT: nonblocking connect retry
[2016/05/23 16:41:30:3272] CLIENT: lws_client_connect_2
[2016/05/23 16:41:30:3272] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:30:3272] CLIENT: nonblocking connect retry
[2016/05/23 16:41:30:8282] CLIENT: lws_client_connect_2
[2016/05/23 16:41:30:8282] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:30:8292] CLIENT: nonblocking connect retry
[2016/05/23 16:41:30:8312] CLIENT: lws_client_connect_2
[2016/05/23 16:41:30:8322] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:30:8332] CLIENT: nonblocking connect retry
[2016/05/23 16:41:31:3342] CLIENT: lws_client_connect_2
[2016/05/23 16:41:31:3342] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:31:3352] CLIENT: nonblocking connect retry
[2016/05/23 16:41:31:3372] CLIENT: lws_client_connect_2
[2016/05/23 16:41:31:3382] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:31:3392] CLIENT: nonblocking connect retry
[2016/05/23 16:41:31:8393] CLIENT: lws_client_connect_2
[2016/05/23 16:41:31:8393] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:31:8403] CLIENT: nonblocking connect retry
[2016/05/23 16:41:31:8413] CLIENT: lws_client_connect_2
[2016/05/23 16:41:31:8423] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:31:8433] CLIENT: nonblocking connect retry
[2016/05/23 16:41:32:3443] CLIENT: lws_client_connect_2
[2016/05/23 16:41:32:3443] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:32:3443] CLIENT: nonblocking connect retry
[2016/05/23 16:41:32:3453] CLIENT: lws_client_connect_2
[2016/05/23 16:41:32:3453] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:32:3453] CLIENT: nonblocking connect retry
[2016/05/23 16:41:32:8463] CLIENT: lws_client_connect_2
[2016/05/23 16:41:32:8463] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:32:8483] CLIENT: nonblocking connect retry
[2016/05/23 16:41:32:8493] CLIENT: lws_client_connect_2
[2016/05/23 16:41:32:8503] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:32:8513] CLIENT: nonblocking connect retry
[2016/05/23 16:41:33:3524] CLIENT: lws_client_connect_2
[2016/05/23 16:41:33:3524] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:33:3534] CLIENT: nonblocking connect retry
[2016/05/23 16:41:33:3534] CLIENT: lws_client_connect_2
[2016/05/23 16:41:33:3544] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:33:3544] CLIENT: nonblocking connect retry
[2016/05/23 16:41:33:8554] CLIENT: lws_client_connect_2
[2016/05/23 16:41:33:8554] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:33:8564] CLIENT: nonblocking connect retry
[2016/05/23 16:41:33:8564] CLIENT: lws_client_connect_2
[2016/05/23 16:41:33:8574] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:33:8574] CLIENT: nonblocking connect retry
[2016/05/23 16:41:34:3594] CLIENT: lws_client_connect_2
[2016/05/23 16:41:34:3594] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:34:3604] CLIENT: nonblocking connect retry
[2016/05/23 16:41:34:3614] CLIENT: lws_client_connect_2
[2016/05/23 16:41:34:3624] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:34:3624] CLIENT: nonblocking connect retry
[2016/05/23 16:41:34:8634] CLIENT: lws_client_connect_2
[2016/05/23 16:41:34:8634] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:34:8644] CLIENT: nonblocking connect retry
[2016/05/23 16:41:34:8654] CLIENT: lws_client_connect_2
[2016/05/23 16:41:34:8654] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:34:8664] CLIENT: nonblocking connect retry
[2016/05/23 16:41:35:3675] CLIENT: lws_client_connect_2
[2016/05/23 16:41:35:3675] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:35:3695] CLIENT: nonblocking connect retry
[2016/05/23 16:41:35:3695] CLIENT: lws_client_connect_2
[2016/05/23 16:41:35:3715] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:35:3725] CLIENT: nonblocking connect retry
[2016/05/23 16:41:35:8725] CLIENT: lws_client_connect_2
[2016/05/23 16:41:35:8725] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:35:8725] CLIENT: nonblocking connect retry
[2016/05/23 16:41:35:8725] CLIENT: lws_client_connect_2
[2016/05/23 16:41:35:8725] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:41:35:8725] CLIENT: nonblocking connect retry

**Comment by andyp1per** [comment]
followed eventually by:

[2016/05/23 16:43:31:3571] NOTICE: wsi 00ACE728: TIMEDOUT WAITING on 2 (did hdr
0, ah 005E9138, wl 0, pfd events 0)
[2016/05/23 16:43:31:3571] INFO: lws_close_free_wsi: real just_kill_connection:
00ACE728
[2016/05/23 16:43:31:3571] INFO: remove_wsi_socket_from_fds: wsi=00ACE728, sock=
380, fds pos=1, end guy pos=2, endfd=0
[2016/05/23 16:43:31:3571] DEBUG: Connection closed before server reply
[2016/05/23 16:43:31:3571] ERR: mirror: LWS_CALLBACK_CLIENT_CONNECTION_ERROR
[2016/05/23 16:43:31:3571] INFO: lws_header_table_detach: wsi 00ACE728: ah 005E9
138 (tsi=0, count = 2)
[2016/05/23 16:43:31:3571] NOTICE: lws_header_table_detach: wsi 00ACE728: ah hel
d 21s, ah.rxpos 0, ah.rxlen 0, mode/state 8 4,wsi->more_rx_waiting 0
[2016/05/23 16:43:31:3571] DEBUG: lws_free_wsi: 00ACE728, remaining wsi 1
[2016/05/23 16:43:31:3581] NOTICE: wsi 00ACCF58: TIMEDOUT WAITING on 2 (did hdr
0, ah 005E84B8, wl 0, pfd events 0)
[2016/05/23 16:43:31:3581] INFO: lws_close_free_wsi: real just_kill_connection:
00ACCF58
[2016/05/23 16:43:31:3581] INFO: remove_wsi_socket_from_fds: wsi=00ACCF58, sock=
332, fds pos=0, end guy pos=1, endfd=380
[2016/05/23 16:43:31:3581] DEBUG: Connection closed before server reply
[2016/05/23 16:43:31:3581] ERR: dumb: LWS_CALLBACK_CLIENT_CONNECTION_ERROR
[2016/05/23 16:43:31:3581] INFO: lws_header_table_detach: wsi 00ACCF58: ah 005E8
4B8 (tsi=0, count = 1)
[2016/05/23 16:43:31:3581] NOTICE: lws_header_table_detach: wsi 00ACCF58: ah hel
d 21s, ah.rxpos 0, ah.rxlen 0, mode/state 8 4,wsi->more_rx_waiting 0
[2016/05/23 16:43:31:3581] DEBUG: lws_free_wsi: 00ACCF58, remaining wsi 0
[2016/05/23 16:43:31:3581] NOTICE: dumb: connecting
[2016/05/23 16:43:31:3581] INFO: lws_header_table_attach: wsi 00ACE728: ah 00000
000 (tsi 0)
[2016/05/23 16:43:31:3581] INFO: lws_header_table_attach: wsi 00ACE728: ah 005E8
4B8: count 1 (on exit)
[2016/05/23 16:43:31:3581] CLIENT: lws_client_connect: direct conn
[2016/05/23 16:43:31:3581] CLIENT: lws_client_connect_2
[2016/05/23 16:43:31:3591] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:43:31:3591] DEBUG: insert_wsi_socket_into_fds: 00ACE728: tsi=0, s
ock=332, pos-in-fds=0
[2016/05/23 16:43:31:3591] CLIENT: nonblocking connect retry
[2016/05/23 16:43:31:3591] NOTICE: mirror: connecting
[2016/05/23 16:43:31:3591] INFO: lws_header_table_attach: wsi 00ACCF58: ah 00000
000 (tsi 0)
[2016/05/23 16:43:31:3591] INFO: lws_header_table_attach: wsi 00ACCF58: ah 005E9
138: count 2 (on exit)
[2016/05/23 16:43:31:3591] CLIENT: lws_client_connect: direct conn
[2016/05/23 16:43:31:3591] CLIENT: lws_client_connect_2
[2016/05/23 16:43:31:3591] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:43:31:3601] DEBUG: insert_wsi_socket_into_fds: 00ACCF58: tsi=0, s
ock=380, pos-in-fds=1
[2016/05/23 16:43:31:3601] CLIENT: nonblocking connect retry
[2016/05/23 16:43:31:3601] CLIENT: lws_client_connect_2
[2016/05/23 16:43:31:3601] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:43:31:3601] CLIENT: nonblocking connect retry
[2016/05/23 16:43:31:3601] CLIENT: lws_client_connect_2
[2016/05/23 16:43:31:3601] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:43:31:3601] CLIENT: nonblocking connect retry
[2016/05/23 16:43:31:8601] CLIENT: lws_client_connect_2
[2016/05/23 16:43:31:8601] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:43:31:8601] CLIENT: nonblocking connect retry
[2016/05/23 16:43:31:8611] CLIENT: lws_client_connect_2
[2016/05/23 16:43:31:8611] CLIENT: lws_client_connect_2: address 127.0.0.1
[2016/05/23 16:43:31:8611] CLIENT: nonblocking connect retry
[2016/05/23 16:43:32:3612] CLIENT: lws_client_connect_2

**Comment by andyp1per** [comment]
Ok, so I think the basic problem is that you have the connect timeouts hardcoded to 20s. I'm guessing that Linux fails in a different way and hence you don't trip up on this. But if I apply the following patch, I can then turn down the timeout to something more reasonable.

diff --git a/lib/client-handshake.c b/lib/client-handshake.c
index 01a74d7..0df66cb 100644
--- a/lib/client-handshake.c
+++ b/lib/client-handshake.c
@@ -178,7 +178,7 @@ lws_client_connect_2(struct lws *wsi)
 					wsi->user_space, NULL, 0);
 		lws_set_timeout(wsi,
 			PENDING_TIMEOUT_AWAITING_CONNECT_RESPONSE,
-							      AWAITING_TIMEOUT);
+				context->timeout_secs);
 #ifdef LWS_USE_IPV6
 		if (LWS_IPV6_ENABLED(context)) {
 			v = (struct sockaddr *)&client_addr6;
@@ -272,7 +272,7 @@ lws_client_connect_2(struct lws *wsi)
 		}
 
 		lws_set_timeout(wsi, PENDING_TIMEOUT_AWAITING_PROXY_RESPONSE,
-				AWAITING_TIMEOUT);
+				context->timeout_secs);
 
 		wsi->mode = LWSCM_WSCL_WAITING_PROXY_REPLY;
 
@@ -290,7 +290,7 @@ lws_client_connect_2(struct lws *wsi)
 	 */
 
 	lws_set_timeout(wsi, PENDING_TIMEOUT_SENT_CLIENT_HANDSHAKE,
-			AWAITING_TIMEOUT);
+			context->timeout_secs);
 
 	wsi->mode = LWSCM_WSCL_ISSUE_HANDSHAKE;
 	pfd.fd = wsi->sock;

Does this seem reasonable to you? If it does I will cook up a pull request.

**Comment by andyp1per** [comment]
FWIW I put some debug in the connect attempt. The error number alternates between
WSAEWOULDBLOCK 10035
and
WSAEALREADY 10037
You don't ever get connection refused, which I would expect on Linux.

**Comment by andyp1per** [comment]
I don't know whether this helps:
http://itamarst.org/writings/win32sockets.html

**Comment by lws-team** [comment]
Thanks... that explains the 20s.  And your link explains windows is really bad at being different for the sake of it

```
Detecting failed connects is done using exception list of select(2)
    On Unix, a failed connect will return an error upon reading or writing from the socket, and select() will mark the socket as notified. The exception list in select(2) is used for detecting OOB data. On Windows, not the read or write lists but rather the exception list is used for detecting failed connects. 
On Windows failed connects, reads and writes still return WSAEWOULDBLOCK
    Reads or writes to the socket will return a WSAEWOULDBLOCK even if connection failed. The errno is available via getsockopt with SO_ERROR.
```

The real problem and fix is detect windows' strange way of signalling connection refused.  In this case it means add something like this...

```
ff --git a/lib/client-handshake.c b/lib/client-handshake.c
index acce53a..4558863 100644
--- a/lib/client-handshake.c
+++ b/lib/client-handshake.c
@@ -208,7 +208,22 @@ lws_client_connect_2(struct lws *wsi)
 #endif
                ) {
                        lwsl_client("nonblocking connect retry\n");
-
+#ifdef _WIN32
+                       {
+                               struct timeval zt;
+                               fd_set fds;
+
+                               zt.tv_sec = 0;
+                               zt.tv_usec = 0;
+
+                               FDS_ZERO(&fds);
+                               FDS_SET(wsi->sock, &fds);
+                               if (select(wsi->sock + 1, NULL, NULL, &fds, &zt) == 1) {
+                                       lwsl_debug("Connection refused\n");
+                                       goto failed;
+                               }
+                       }
+#endif
                        /*
                         * must do specifically a POLLOUT poll to hear
                         * about the connect completion
```

Can you give it a try?  It may need some meddling...

**Comment by andyp1per** [comment]
I think you mean FD_ZERO and FD_SET, but that unfortunately makes no difference. If I put a breakpoint  on the debug statement I never hit it.

**Comment by lws-team** [comment]
Well, it follows the recipe from the twisted guy afaict.  If that doesn't actually detect it I dunno.

Overall it seems lws acts well and catches this with a timeout; it's not blocking.  This connect action is not really the same as a normal timeout since it usually involves a dns lookup, which can be slow... 20s sounds reasonable.

The only problem is still detecting connection refused so we fail quickly I think.

**Comment by andyp1per** [comment]
So hardcoded timeouts still suck :). Any objection to me introducing a connection_timeout in the lws context?
I'm on OSX today, so won't be able to dig into this until next week now.

**Comment by lws-team** [comment]
Well, they don't suck as bad as following a general timeout down to 1s as you tried, when the dns arrangements take longer than that...

I think it's fine to allow setting it, so long as 0 == 20 so nothing changes for people who don't care (ie, anyone on unix or bsd / mac ).

**Comment by andyp1per** [comment]
This appears to work:
'''
diff --git a/lib/client-handshake.c b/lib/client-handshake.c
index 01a74d7..3d772de 100644
--- a/lib/client-handshake.c
+++ b/lib/client-handshake.c
@@ -230,6 +230,21 @@ lws_client_connect_2(struct lws *wsi)
 #endif
                ) {
                        lwsl_client("nonblocking connect retry\n");
+#ifdef _WIN32
+                       {
+                               int optVal;
+                               int optLen = sizeof(int);
+                               if (getsockopt(wsi->sock, SOL_SOCKET, SO_ERROR, (char*)&optVal, &optLen)
+                                       != SOCKET_ERROR && optVal != 0
+                                       && optVal != LWS_EALREADY
+                                       && optVal != LWS_EINPROGRESS
+                                       && optVal != LWS_EWOULDBLOCK
+                                       && optVal != WSAEINVAL) {
+                                       lwsl_debug("Connect failed SO_ERROR=%d\n", optVal);
+                                       goto failed;
+                               }
+                       }
+#endif

                        /*
                         * must do specifically a POLLOUT poll to hear
'''

**Comment by andyp1per** [comment]
Ok, need your help. This doesn't seem to be quite right - sometimes it catches the error sometimes it doesn't. I think this is down to other parts of the code resetting the event and error and so we miss it. I think the correct way to listen for this given the calls made is to look out for FD_CONNECT and if you see that with an error then bail: (see https://msdn.microsoft.com/en-us/library/windows/desktop/ms741576(v=vs.85).aspx for example). I have the following patch which correctly catches the error but does not seem to cause the event loop to bail. I'm not really following the POLLIN/POLLOUT stuff well so I am sure that I have missed something. Maybe you could take a look and give me your thoughts?

```
diff --git a/lib/client-handshake.c b/lib/client-handshake.c
index 01a74d7..709aacb 100644
--- a/lib/client-handshake.c
+++ b/lib/client-handshake.c
@@ -229,7 +229,22 @@ lws_client_connect_2(struct lws *wsi)
                        || LWS_ERRNO == WSAEINVAL
 #endif
                ) {
-                       lwsl_client("nonblocking connect retry\n");
+                       lwsl_client("nonblocking connect retry, errno=%d\n", LWS_ERRNO);
+#ifdef _WIN32
+                       {
+                               int optVal;
+                               int optLen = sizeof(int);
+                               if (getsockopt(wsi->sock, SOL_SOCKET, SO_ERROR, (char*)&optVal, &optLen)
+                                       != SOCKET_ERROR && optVal != 0
+                                       && optVal != LWS_EALREADY
+                                       && optVal != LWS_EINPROGRESS
+                                       && optVal != LWS_EWOULDBLOCK
+                                       && optVal != WSAEINVAL) {
+                                       lwsl_debug("Connect failed SO_ERROR=%d\n", optVal);
+                                       goto failed;
+                               }
+                       }
+#endif

                        /*
                         * must do specifically a POLLOUT poll to hear
diff --git a/lib/lws-plat-win.c b/lib/lws-plat-win.c
index 2c0213a..64c29ca 100644
--- a/lib/lws-plat-win.c
+++ b/lib/lws-plat-win.c
@@ -229,6 +229,18 @@ lws_plat_service_tsi(struct lws_context *context, int timeout_ms, int tsi)
                return -1;
        }

+       if (networkevents.lNetworkEvents & FD_CONNECT) {
+               if (networkevents.iErrorCode[FD_CONNECT_BIT] != 0) {
+                       WSAResetEvent(pt->events[0]);
+                       lwsl_err("Unable to connect errno=%d\n", networkevents.iErrorCode[FD_CONNECT_BIT]);
+                       return -1;
+               }
+               else {
+                       WSAResetEvent(pt->events[0]);
+                       return 0;
+               }
+       }
+
        pfd->revents = (short)networkevents.lNetworkEvents;

        if (pfd->revents & LWS_POLLOUT) {
@@ -404,7 +416,7 @@ lws_plat_insert_socket_into_fds(struct lws_context *context, struct lws *wsi)
        pt->fds[pt->fds_count++].revents = 0;
        pt->events[pt->fds_count] = WSACreateEvent();
        WSAEventSelect(wsi->sock, pt->events[pt->fds_count],
-                      LWS_POLLIN | LWS_POLLHUP);
+                      LWS_POLLIN | LWS_POLLHUP | FD_CONNECT);
 }

 LWS_VISIBLE void
@@ -427,7 +439,7 @@ lws_plat_change_pollfd(struct lws_context *context,
                      struct lws *wsi, struct lws_pollfd *pfd)
 {
        struct lws_context_per_thread *pt = &context->pt[(int)wsi->tsi];
-       long networkevents = LWS_POLLHUP;
+       long networkevents = LWS_POLLHUP | FD_CONNECT;

        if ((pfd->events & LWS_POLLIN))
                networkevents |= LWS_POLLIN;
```

**Comment by andyp1per** [comment]
So I'm running with this change and it appears to work pretty well. Your thoughts?
```
diff --git a/lib/client-handshake.c b/lib/client-handshake.c
index 01a74d7..709aacb 100644
--- a/lib/client-handshake.c
+++ b/lib/client-handshake.c
@@ -229,7 +229,22 @@ lws_client_connect_2(struct lws *wsi)
                        || LWS_ERRNO == WSAEINVAL
 #endif
                ) {
-                       lwsl_client("nonblocking connect retry\n");
+                       lwsl_client("nonblocking connect retry, errno=%d\n", LWS_ERRNO);
+#ifdef _WIN32
+                       {
+                               int optVal;
+                               int optLen = sizeof(int);
+                               if (getsockopt(wsi->sock, SOL_SOCKET, SO_ERROR, (char*)&optVal, &optLen)
+                                       != SOCKET_ERROR && optVal != 0
+                                       && optVal != LWS_EALREADY
+                                       && optVal != LWS_EINPROGRESS
+                                       && optVal != LWS_EWOULDBLOCK
+                                       && optVal != WSAEINVAL) {
+                                       lwsl_debug("Connect failed SO_ERROR=%d\n", optVal);
+                                       goto failed;
+                               }
+                       }
+#endif

                        /*
                         * must do specifically a POLLOUT poll to hear
diff --git a/lib/lws-plat-win.c b/lib/lws-plat-win.c
index 2c0213a..1bf76c7 100644
--- a/lib/lws-plat-win.c
+++ b/lib/lws-plat-win.c
@@ -229,7 +229,19 @@ lws_plat_service_tsi(struct lws_context *context, int timeout_ms, int tsi)
                return -1;
        }

-       pfd->revents = (short)networkevents.lNetworkEvents;
+       if ((networkevents.lNetworkEvents & FD_CONNECT)
+               && networkevents.iErrorCode[FD_CONNECT_BIT] != 0
+               && networkevents.iErrorCode[FD_CONNECT_BIT] != LWS_EALREADY
+               && networkevents.iErrorCode[FD_CONNECT_BIT] != LWS_EINPROGRESS
+               && networkevents.iErrorCode[FD_CONNECT_BIT] != LWS_EWOULDBLOCK
+               && networkevents.iErrorCode[FD_CONNECT_BIT] != WSAEINVAL) {
+               lwsl_debug("Unable to connect errno=%d\n", networkevents.iErrorCode[FD_CONNECT_BIT]);
+               pfd->revents = LWS_POLLHUP;
+       }
+       else {
+               pfd->revents = (short)networkevents.lNetworkEvents;
+       }
+

        if (pfd->revents & LWS_POLLOUT) {
                wsi = wsi_from_fd(context, pfd->fd);
@@ -404,7 +416,7 @@ lws_plat_insert_socket_into_fds(struct lws_context *context, struct lws *wsi)
        pt->fds[pt->fds_count++].revents = 0;
        pt->events[pt->fds_count] = WSACreateEvent();
        WSAEventSelect(wsi->sock, pt->events[pt->fds_count],
-                      LWS_POLLIN | LWS_POLLHUP);
+                      LWS_POLLIN | LWS_POLLHUP | FD_CONNECT);
 }

 LWS_VISIBLE void
@@ -427,7 +439,7 @@ lws_plat_change_pollfd(struct lws_context *context,
                      struct lws *wsi, struct lws_pollfd *pfd)
 {
        struct lws_context_per_thread *pt = &context->pt[(int)wsi->tsi];
-       long networkevents = LWS_POLLHUP;
+       long networkevents = LWS_POLLHUP | FD_CONNECT;

        if ((pfd->events & LWS_POLLIN))
                networkevents |= LWS_POLLIN;
```

**Comment by lws-team** [comment]
Looks very reasonable to me thanks (if I squint at how unreasonable windows is acting differently in the first place).  I moved the client-handshake bit into its own plat_ api to avoid more #ifdefs and pushed it on master, please see if I mangled anything.

https://github.com/warmcat/libwebsockets/commit/b37bd5e5c76ddd6b03647c4440ddfef790839e58

**Comment by andyp1per** [comment]
Ah, you've committed my first fix, which had issues. My second fix is much better and I think you should go with that. My only question was whether triggering POLLHUP in this way was reasonable.

**Comment by lws-team** [comment]
Hm?  The patch is based on your last diff, did I miss something?

**Comment by andyp1per** [comment]
the change to lws-plat-win.c is different in my  last change

**Comment by lws-team** [comment]
I see... unfortunately the diff you pasted wouldn't apply, so I had to add each part by hand, it looks like I grabbed that bit from the previous version.

How about this?  https://github.com/warmcat/libwebsockets/commit/6c5fe726a52946b18d0a67b231a35d5188c95e35

**Comment by andyp1per** [comment]
yes looks right thanks!

**Comment by lws-team** [comment]
Thanks... if there's something else to do / fix please reopen.

**Comment by andyp1per** [comment]
Any chance this could get into the 2.0 branch?

**Comment by lws-team** [comment]
Should be done.

**Comment by andyp1per** [comment]
Thanks!
